elfi/examples/lotka_volterra.py

```diff
@@ -0,0 +1,253 @@
+"""Example of inference with the stochastic Lotka-Volterra predator prey model.
+
+Treatment roughly follows:
+Owen, J., Wilkinson, D. and Gillespie, C. (2015) Likelihood free inference for
+   Markov processes: a comparison. Statistical Applications in Genetics and
+   Molecular Biology, 14(2).
+"""
+
+import logging
+from functools import partial
+
+import numpy as np
+import scipy.stats as ss
+
+import elfi
+
+
+def lotka_volterra(r1, r2, r3, prey_init=50, predator_init=100, sigma=0., n_obs=16, time_end=30.,
+                   batch_size=1, random_state=None, return_full=False):
+    r"""Generate sequences from the stochastic Lotka-Volterra model.
+
+    The Lotka-Volterra model is described by 3 reactions
+
+    R1 : X1 -> 2X1       # prey reproduction
+    R2 : X1 + X2 -> 2X2  # predator hunts prey and reproduces
+    R3 : X2 -> 0         # predator death
+
+    The system is solved using the Direct method.
+
+    Gillespie, D. T. (1977) Exact stochastic simulation of coupled chemical reactions.
+        The Journal of Physical Chemistry 81 (25), 2340–2361.
+    Lotka, A. J. (1925) Elements of physical biology. Williams & Wilkins Baltimore.
+    Volterra, V. (1926) Fluctuations in the abundance of a species considered mathematically.
+        Nature 118, 558–560.
+
+
+    Parameters
+    ----------
+    r1 : float or np.array
+        Rate of R1.
+    r2 : float or np.array
+        Rate of R2.
+    r3 : float or np.array
+        Rate of R3.
+    prey_init : int or np.array, optional
+        Initial number of prey.
+    predator_init : int or np.array, optional
+        Initial number of predators.
+    sigma : float or np.array, optional
+        Standard deviation of the Gaussian noise added to measurements.
+    n_obs : int, optional
+        Number of observations to return at integer frequency.
+    time_end : float, optional
+        Time allowed for reactions in the Direct method (not the wall time).
+    batch_size : int, optional
+    random_state : np.random.RandomState, optional
+    return_full : bool, optional
+        If True, return a tuple (observed_stock, observed_times, full_stock, full_times).
+
+    Returns
+    -------
+    stock_obs : np.array
+        Observations in shape (batch_size, n_obs, 2).
+
+    """
+    random_state = random_state or np.random
+
+    r1 = np.asanyarray(r1).reshape(-1)
+    r2 = np.asanyarray(r2).reshape(-1)
+    r3 = np.asanyarray(r3).reshape(-1)
+    prey_init = np.asanyarray(prey_init).reshape(-1)
+    predator_init = np.asanyarray(predator_init).reshape(-1)
+    sigma = np.asanyarray(sigma).reshape(-1)
+
+    n_full = 20000
+    stock = np.empty((batch_size, n_full, 2), dtype=np.int32)
+    stock[:, 0, 0] = prey_init
+    stock[:, 0, 1] = predator_init
+    stoichiometry = np.array([[1, 0], [-1, 1], [0, -1], [0, 0]], dtype=np.int32)
+    times = np.empty((batch_size, n_full))
+    times[:, 0] = 0
+    ii = 0
```

Comment:
Could be moved to just before the start of the while loop to indicate that `ii` is auxiliary